### PR TITLE
Add HEXL-FPGA compatibility for DyadicMultiply and KeySwitch

### PR DESCRIPTION
**Pull request by tgonzalez89-intel**
Added an internal implementation for DyadicMultiply and KeySwitch and a CMake option and compiler flag to disable the normal DyadicMultiply and KeySwitch and only leave the internal one available. This makes HEXL compatible with HEXL-FPGA. With this both libraries can use HEXL's header but HEXL-FPGA can override DyadicMultiply and KeySwitch with its own implementation while leaving the internal implementation as a fallback if the user desires to use the CPU implementation.

Renamed all instances of CKKS in all functions and file names.

Added a new root_of_unity_powers_ptr option to KeySwitch. This option will be used in HEXL-FPGA.

### CODE REVIEW

**Review by hamishun — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -73,10 +74,15 @@ message(STATUS "HEXL_EXPERIMENTAL:             ${HEXL_EXPERIMENTAL}")
 message(STATUS "HEXL_SHARED_LIB:               ${HEXL_SHARED_LIB}")
 message(STATUS "HEXL_TESTING:                  ${HEXL_TESTING}")
 message(STATUS "HEXL_TREAT_WARNING_AS_ERROR:   ${HEXL_TREAT_WARNING_AS_ERROR}")
+message(STATUS "HEXL_FPGA_COMPAT:              ${HEXL_FPGA_COMPAT}")
```
Comment (joserochh) [inline CMakeLists.txt:77]: Maybe we can use HEXL_FPGA_COMPATIBLE instead? For explicitness.

File: hexl/experimental/seal/key-switch-internal.cpp [hunk 2]
```diff
@@ -16,13 +19,20 @@
...
+void KeySwitch(uint64_t* result, const uint64_t* t_target_iter_ptr, uint64_t n,
+               uint64_t decomp_modulus_size, uint64_t key_modulus_size,
+               uint64_t rns_modulus_size, uint64_t key_component_count,
+               const uint64_t* moduli, const uint64_t** k_switch_keys,
+               const uint64_t* modswitch_factors,
+               const uint64_t* root_of_unity_powers_ptr) {
+  if (root_of_unity_powers_ptr != nullptr) {
+    std::cout << "Passing root_of_unity_powers_ptr as a parameter is not "
+                 "supported yet.\n";
+    assert(root_of_unity_powers_ptr == nullptr);
```
Comment (joserochh) [inline hexl/experimental/seal/key-switch-internal.cpp:33]: Is it possible to change this to throw an exception error?

File: CMakeLists.txt [hunk 3]
```diff
@@ -73,10 +74,15 @@ message(STATUS "HEXL_EXPERIMENTAL:             ${HEXL_EXPERIMENTAL}")
...
 
 hexl_check_compiler_version()
 hexl_add_compiler_definition()
 
+if (HEXL_FPGA_COMPAT)
+  add_definitions(-DHEXL_FPGA_COMPAT)
```
Comment (jlhcrawford) [inline CMakeLists.txt:83]: This function has been superseded by `add_compile_definitions`
```suggestion
  add_compile_definitions(HEXL_FPGA_COMPAT)
```

File: hexl/include/hexl/experimental/seal/key-switch-internal.hpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+/// decomp_modulus_size entries, each with
+/// coeff_count * ((key_modulus_size - 1)+ (key_component_count - 1) *
+/// (key_modulus_size) + 1) entries
+/// @param[in] modswitch_factors Array of modulus switch factors
+void KeySwitch(uint64_t* result, const uint64_t* t_target_iter_ptr, uint64_t n,
+               uint64_t decomp_modulus_size, uint64_t key_modulus_size,
+               uint64_t rns_modulus_size, uint64_t key_component_count,
+               const uint64_t* moduli, const uint64_t** k_switch_keys,
+               const uint64_t* modswitch_factors,
+               const uint64_t* root_of_unity_powers_ptr = nullptr);
```
Comment (jlhcrawford) [inline hexl/include/hexl/experimental/seal/key-switch-internal.hpp:37]: Is there a reason why `root_of_unity_powers_ptr` is not documented in the Doxygen above?

**Review by jlhcrawford — CHANGES_REQUESTED** [review]

**Comment by ymeng-git** [comment]
continue as PR 109.

**Pull request closed without merging**